### PR TITLE
Add %%bq query to build BigQuery query objects incrementally

### PR DESCRIPTION
**Pull request by yebrahim**
This PR adds the `%%bq query` which can be used to build BigQuery query modules that can consume other, similar modules, or UDF objects.

The change also adds recursive subquery expansion. It does *not* cover circular subquery reference, this is rather delegated to the BigQuery service.

### CODE REVIEW

File: google/datalab/bigquery/_query.py [hunk 1]
```diff
@@ -91,6 +107,41 @@ def __init__(self, sql, context=None, values=None, udfs=None, data_sources=None,
           raise Exception('Referenced external table %s has no known schema' % name)
         self._external_tables[name] = table._to_query_json()
 
+  @property
+  def expanded_sql(self):
```
Comment (ojarjur) [inline google/datalab/bigquery/_query.py]: We really need to add a unit test for this.

Comment (yebrahim) [inline google/datalab/bigquery/_query.py]: I agree, we have a lot of work to do on unit tests.

File: google/datalab/bigquery/_query.py [hunk 2]
```diff
@@ -65,20 +67,34 @@ def __init__(self, sql, context=None, values=None, udfs=None, data_sources=None,
...
 
     if data_sources is None:
       data_sources = {}
 
     self._code = None
     self._imports = []
-    if values is None:
-      values = kwargs
+    if self._values is None:
+      self._values = kwargs
```
Comment (chmeyers) [inline google/datalab/bigquery/_query.py]: This is going to conflict with my change. :)

Just make it self._values = {} and remove kwargs above if you want to reduce the conflict.

File: google/datalab/bigquery/_query.py [hunk 3]
```diff
@@ -91,13 +107,52 @@ def __init__(self, sql, context=None, values=None, udfs=None, data_sources=None,
...
+    def _recurse_subqueries(query):
+      """Recursively scan subqueries and add their pieces to global scope udfs and subqueries
+      """
+      if query._subqueries:
+        subqueries.update(query._subqueries)
+      if query._udfs:
+        udfs.update(set(query._udfs))
+      if query._subqueries:
+        for subquery in query._subqueries:
+          subquery = self._values[subquery]
```
Comment (chmeyers) [inline google/datalab/bigquery/_query.py]: Why not just _recurse_subqueries(self._values[subquery]) ?

File: google/datalab/bigquery/_query.py (hunk 2 above)
Comment (yebrahim) [inline google/datalab/bigquery/_query.py]: Done

File: google/datalab/bigquery/_query.py (hunk 3 above)
Comment (yebrahim) [inline google/datalab/bigquery/_query.py]: Done

**Pull request merged**